### PR TITLE
[wip] Implement hierarchical progress bar control in huggingface_hub

### PR DESCRIPTION
**Pull request by lappemic**
As discussed in #1595, i worked on enabling/disabling specific progress bars. 

Changes:
- Added private utility functions `_set_progress_bar_state` and `_get_progress_bar_state` to the `tqdm` module for hierarchical state management via the `progress_bar_states` dict.
- Updated `disable_progress_bars`, `enable_progress_bars`, and `are_progress_bars_disabled` to utilize this hierarchical approach.

Marked as [WIP] since this is an initial draft for discussion. I'm unsure about the implementation of the hierarchical logic and look forward to your feedback, @Wauplin.

### CODE REVIEW

File: src/huggingface_hub/utils/tqdm.py [hunk 1]
```diff
@@ -66,58 +66,124 @@
...
+# Essentially, the environment variable's setting overrides any code-based configurations.
 #
-# By default, progress bars are enabled.
-_hf_hub_progress_bars_disabled: bool = HF_HUB_DISABLE_PROGRESS_BARS or False
+# If `HF_HUB_DISABLE_PROGRESS_BARS` is not defined (None), it implies that users can manage
+# progress bar visibility through code. By default, progress bars are turned on.
 
 
-def disable_progress_bars() -> None:
+progress_bar_states = {}
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply src/huggingface_hub/utils/tqdm.py:78]: An alternative approach would be to have a flat dictionary instead (typed as `Dict[str, bool]`. So something like:

```py
{"foo.bar": False}
```

When running `enable_progress_bars("foo.bar.something")`, it becomes

```py
{"foo.bar": False, "foo.bar.something": True}
```

Then if you run `disable_progress_bars("foo")`, it becomes:
```py
{"foo": False}
```

---

The logic to do that would be:
- for `disable_progress_bars(name)`: loop through all entries. If `key.startswith(f"{name}.")` => delete the entry. Then set `states[name]` to False
- for `enable_progress_bars(name)`: loop through all entries. If `key.startswith(f"{name}.")`  => delete the entry. Then set `states[name]` to True
- for `is_progress_bars_enabled(name)`: check for `name` => if present, return value. If not, check for "name minus the last part". Then "name minus the 2 last parts", etc until you reach `""`.

WDYT?

[the remark this answers is not in the document]
Comment (lappemic) [reply src/huggingface_hub/utils/tqdm.py:78]: Thank you so much for your inputs @Wauplin! 🙏 The suggestion to have a flat dict made it easier for me to grasp! I now also removed the private functions and implemented the simplified logic right in the corresponding functions. I think this might work now. 😄

**Comment by lappemic** [comment]
Additionally to the simplified logic i tried to improve the docstrings according to the example in #2209 and added an example for the hierarchical control. I hope this helps the clearity.

Again, thank you very much @Wauplin for the guidance on the logic implementation, i really appreciate you taking time for this!

File: src/huggingface_hub/utils/tqdm.py [hunk 2]
```diff
@@ -53,71 +53,141 @@
...
+Hierarchical Control Example:
+    ```python
+    # Disable progress bars for a specific group
+    disable_progress_bars("peft.foo")
+
+    # Check state of different groups
+    print(are_progress_bars_disabled("peft"))         # Outputs: False
+    print(are_progress_bars_disabled("peft.something"))  # Outputs: False
+    print(are_progress_bars_disabled("peft.foo"))       # Outputs: True
+    print(are_progress_bars_disabled("peft.foo.bar"))   # Outputs: True
```
Comment (Wauplin) [inline src/huggingface_hub/utils/tqdm.py:66]: Usually the preferred way to document inputs and outputs is to make it look like you are running in a terminal. It renders better in the docs.

```suggestion
    ```python
    # Disable progress bars for a specific group
    >>> disable_progress_bars("peft.foo")

    # Check state of different groups
    >>> are_progress_bars_disabled("peft"))
    False
    >>> are_progress_bars_disabled("peft.something")
    False
    >>> are_progress_bars_disabled("peft.foo"))
    True
    >>> are_progress_bars_disabled("peft.foo.bar"))
    True
```

(to be adapted in the example below as well)

**Review by Wauplin — COMMENTED** [review]
Hi @lappemic thanks a lot for the changes! I haven't tested it myself yet but it looks much cleaner like this :)

>  I think this might work now. 😄

Best way to be sure will be to implement tests! You can do that [in this file](https://github.com/huggingface/huggingface_hub/blob/main/tests/test_utils_tqdm.py). I'll let you have a look at it but it should be quite straightforward to add them by looking at the existing ones. Make sure to test every possible corner cases (potentially in separate test methods).

And finally, to make this PR complete you will have to update the documentation [in this file](https://github.com/huggingface/huggingface_hub/blob/main/docs/source/en/package_reference/utilities.md#configure-progress-bars) (which is rendered [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_2217/en/package_reference/utilities#configure-progress-bars)). 

Thanks to you for taking time to implement that!

File: src/huggingface_hub/utils/tqdm.py (hunk 2 above)
Comment (lappemic) [reply src/huggingface_hub/utils/tqdm.py]: Adapted 👍

**Comment by lappemic** [comment]
> Hi @lappemic thanks a lot for the changes! I haven't tested it myself yet but it looks much cleaner like this :)
Thanks @Wauplin, this is good to hear. 👍 

> Best way to be sure will be to implement tests! 
Absoultely. I haven't thought about it then. Implemented them now. I hope i got all of the corner cases. It feels like an extensive test at least. (Not really experienced with testing, so happy for every input as well)

> And finally, to make this PR complete you will have to update the documentation.
I updated the documantation. About the wording and example style i was not sure. Just took the one you used in the original issue to showcase it. Let me know if i can adjust something there.

**Review by Wauplin — APPROVED** [review]
Thanks @lappemic! Finally took the time to review this PR and everything looks ready to merge for me :) :tada: 

I took the liberty to push some changes to your PR. Mostly irrelevant changes except:
- https://github.com/huggingface/huggingface_hub/pull/2217/commits/078b8d97ce04f1698374febaceea3a51247462f5 => we forgot to respect the `name` parameter in `tqdm` which makes the PR useless otherwise :smile: Fixed this, now possible to do `for _ in tqdm(range(10), name="foo.bar"): ...`
- https://github.com/huggingface/huggingface_hub/pull/2217/commits/7266961f64c0f2ea24ee1571b77d8d9f903fa338 => add groups where `huggingface_hub` uses tqdm :)

This is now ready to be merged. Will do once the CI is green and then make a release very soon!

**Comment by lappemic** [comment]
Perfect!

Thanks a lot for the clean up and final commiits! 🙏
